File: fuzz.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <random>


using namespace std;

struct Rule{
    string l;
    string r; 
    Rule(const string& left, const string& right) : l(left), r(right) {}
};


vector<Rule> allRulesT2;
vector<Rule> allRulesT;

string alphabet = "ab";

void initRulesT(){
    allRulesT.push_back(Rule("aaaa", "a"));
    allRulesT.push_back(Rule("aaab", "b"));
    allRulesT.push_back(Rule("bbba", "ba"));
    allRulesT.push_back(Rule("bbbb", "bb"));
    allRulesT.push_back(Rule("ababb", "babb"));
    allRulesT.push_back(Rule("baaba", "bba"));
    allRulesT.push_back(Rule("baabb", "bbb"));
    allRulesT.push_back(Rule("bbaaa", "baab"));
    allRulesT.push_back(Rule("bbaab", "baa"));
    allRulesT.push_back(Rule("bbabb", "abab"));
    allRulesT.push_back(Rule("baba", "baa"));
    allRulesT.push_back(Rule("babbaa", "babba"));
    allRulesT.push_back(Rule("babbab", "abb"));
}

void initRulesT2(){
    /*
    allRulesT2.push_back(Rule("aaaa", "a"));
    allRulesT2.push_back(Rule("aaab", "b"));
    allRulesT2.push_back(Rule("bbaaa", "baab"));
    allRulesT2.push_back(Rule("bbaab", "baa"));
    allRulesT2.push_back(Rule("bbabb", "abab"));
    allRulesT2.push_back(Rule("babbaa", "babba"));
    allRulesT2.push_back(Rule("bab", "abb"));
    allRulesT2.push_back(Rule("baa", "abb"));
    allRulesT2.push_back(Rule("abb", "bb"));
    allRulesT2.push_back(Rule("bba", "bb"));
    allRulesT2.push_back(Rule("bb", "ba"));
    allRulesT2.push_back(Rule("aba", "ba"));
    */
    allRulesT2.push_back(Rule("aaaa", "a"));
    allRulesT2.push_back(Rule("aaab", "b"));
    allRulesT2.push_back(Rule("aaaba", "babb"));
    allRulesT2.push_back(Rule("baba", "baab"));
    allRulesT2.push_back(Rule("bab", "baa"));
    allRulesT2.push_back(Rule("babbb", "babba"));
    allRulesT2.push_back(Rule("aaba", "baa"));
    allRulesT2.push_back(Rule("baa", "abb"));
    allRulesT2.push_back(Rule("aba", "bb"));
    allRulesT2.push_back(Rule("bb", "ba"));
    
}

// l - левая граница подстроки в s (индекс), которую нужно заменить по правилу
void executeRule(Rule rule, string& s, int l){
    //string& replace (size_t pos, size_t len, const string& str);
    int len = rule.l.length();
    s.replace(l, len, rule.r);
}

//проверим, можно ли применить правило из списка Rules
pair<bool, int> checkIsRuleMayBeApplied(Rule rule, const string& s){
    pair<bool, int> p;
    size_t found = s.find(rule.l);
    p.second = found;

    if (found != std::string::npos) {
        p.first = true;
        return p;
    }
    
    p.first = false;
    return p;
}


//приводим строку к нормальной форме
void normalize(string& s){
    for (Rule rule: allRulesT2){
        pair<bool, int> p = checkIsRuleMayBeApplied(rule, s);
        if (p.first){
            executeRule(rule, s, p.second);
            normalize(s);
        }else{
            continue;
        }
    }
}


//генерация случайного номера правила в системе T
// n - сколько всего правил
int getRandomRuleT(int n) {
    static std::random_device rd;
    static std::mt19937 gen(rd());
    static std::uniform_int_distribution<int> dist(0, n - 1);
    return dist(gen);
}

//генерация случайного номера правила в системе T2
// n - сколько всего правил
int getRandomRuleT2(int n) {
    static std::random_device rd;
    static std::mt19937 gen(rd());
    static std::uniform_int_distribution<int> dist(0, n - 1);
    return dist(gen);
}

//генерация случайной строки, заданной длины n, состоящей из символов 
//фиксированного алфавита
string getRandomString(int n, const string& alphabet) {
    string s;
    static std::random_device rd;
    static std::mt19937 gen(rd());
    static std::uniform_int_distribution<int> dist(0, alphabet.size() - 1);
    for (int i = 0; i < n; i++){
        s += alphabet[dist(gen)];
    }
    return s;
}

// генерируем случайные 7 правил и посмотрим, в какую строку перейдём
//при такой цепочке правил из произвольно выбранной строки в системе T
pair<string, vector<int>> T(string s){
    vector<int> path;
    int n = allRulesT.size();
    for (int i = 0; i < 7; i++){
        int ruleNum = getRandomRuleT(n);
        Rule rule = allRulesT[ruleNum];
        pair<bool, int> p = checkIsRuleMayBeApplied(rule, s);
        path.push_back(ruleNum);
        if (p.first == true){
            executeRule(rule, s, p.second);
        }
    }
    
    pair<string, vector<int>> p;
    p.first = s;
    p.second = path;
    return p;
}


//будем генерировать 50000 тестов
//вернём true, если все тесты успешны,
//иначе false
//возвращаем строку, на которой была ошибка
//и последовательность правил
//ВАЖНО: 
//выбранную строку и строку, которая получится в исходной системе,
//доведём в новой системе до нормальных форм
//они должны совпасть 
pair<bool, pair<string, vector<int>>> fuzzing(){
    pair<bool, pair<string, vector<int>>> p1;
    p1.first = true;
    p1.second.first = "";
    vector<int> emptyVector;
    p1.second.second = emptyVector;
    for (int i = 0; i < 50000; i++){
        string s = getRandomString(15, alphabet);
        pair<string, vector<int>> p2 = T(s);
        string s2 = p2.first;
        vector<int> path = p2.second;
        normalize(s);
        normalize(s2);
        if (s != s2){
            p1.first = false;
            p1.second.first = s;
            p1.second.second = path;
            return p1;
        }
    }
    return p1;
}

int main(){
    initRulesT();
    initRulesT2();
    pair<bool, pair<string, vector<int>>> p = fuzzing();
    if (p.first){
        cout << "все тесты успешны" << endl;
    }else{
        cout << "найдена ошибка: " << p.second.first << " и правила:" << endl;
        for (int ruleNum: p.second.second){
            cout << ruleNum << endl;
        }
    }
}

File: fuzz.rs
use std::time::{SystemTime, UNIX_EPOCH};

struct Rule {
    l: String,
    r: String,
}

impl Rule {
    fn new(left: &str, right: &str) -> Self {
        Self {
            l: left.to_string(),
            r: right.to_string(),
        }
    }
}

// Простой самодельный RNG
struct SimpleRng {
    state: u64,
}

impl SimpleRng {
    fn new() -> Self {
        let seed = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_nanos() as u64;
        Self { state: seed }
    }
    
    fn gen_range(&mut self, range: std::ops::Range<usize>) -> usize {
        self.state = self.state.wrapping_mul(6364136223846793005).wrapping_add(1);
        let start = range.start;
        let end = range.end;
        start + (self.state as usize) % (end - start)
    }
}

static mut ALL_RULES_T2: Vec<Rule> = Vec::new();
static mut ALL_RULES_T: Vec<Rule> = Vec::new();
const ALPHABET: &str = "ab";

fn init_rules_t() {
    unsafe {
        ALL_RULES_T.push(Rule::new("aaaa", "a"));
        ALL_RULES_T.push(Rule::new("aaab", "b"));
        ALL_RULES_T.push(Rule::new("bbba", "ba"));
        ALL_RULES_T.push(Rule::new("bbbb", "bb"));
        ALL_RULES_T.push(Rule::new("ababb", "babb"));
        ALL_RULES_T.push(Rule::new("baaba", "bba"));
        ALL_RULES_T.push(Rule::new("baabb", "bbb"));
        ALL_RULES_T.push(Rule::new("bbaaa", "baab"));
        ALL_RULES_T.push(Rule::new("bbaab", "baa"));
        ALL_RULES_T.push(Rule::new("bbabb", "abab"));
        ALL_RULES_T.push(Rule::new("baba", "baa"));
        ALL_RULES_T.push(Rule::new("babbaa", "babba"));
        ALL_RULES_T.push(Rule::new("babbab", "abb"));
    }
}

fn init_rules_t2() {
    unsafe {
        ALL_RULES_T2.push(Rule::new("bb", "ba"));
        ALL_RULES_T2.push(Rule::new("aba", "ba"));
        ALL_RULES_T2.push(Rule::new("baa", "ba"));
        ALL_RULES_T2.push(Rule::new("bab", "ba"));
        ALL_RULES_T2.push(Rule::new("aaaa", "a"));
        ALL_RULES_T2.push(Rule::new("aaab", "b"));
    }
}

fn execute_rule(rule: &Rule, s: &mut String, l: usize) {
    let len = rule.l.len();
    s.replace_range(l..l + len, &rule.r);
}

fn check_is_rule_may_be_applied(rule: &Rule, s: &str) -> (bool, usize) {
    if let Some(found) = s.find(&rule.l) {
        (true, found)
    } else {
        (false, 0)
    }
}

fn normalize(s: &mut String) {
    unsafe {
        for rule in &ALL_RULES_T2 {
            let (can_apply, pos) = check_is_rule_may_be_applied(rule, s);
            if can_apply {
                execute_rule(rule, s, pos);
                normalize(s);
            }
        }
    }
}

fn get_random_rule_t() -> usize {
    let mut rng = SimpleRng::new();
    unsafe {
        rng.gen_range(0..ALL_RULES_T.len())
    }
}

fn get_random_string(n: usize) -> String {
    let mut rng = SimpleRng::new();
    let alphabet: Vec<char> = ALPHABET.chars().collect();
    (0..n)
        .map(|_| {
            let idx = rng.gen_range(0..alphabet.len());
            alphabet[idx]
        })
        .collect()
}

// генерируем случайные 7 правил и посмотрим, в какую строку перейдём
//при такой цепочке правил из произвольно выбранной строки в системе T
fn t(s: &str) -> (String, Vec<usize>) {
    let mut current = s.to_string();
    let mut path = Vec::new();

    for _ in 0..7 {
        let rule_num = get_random_rule_t();
        unsafe {
            let rule = &ALL_RULES_T[rule_num];
            let (can_apply, pos) = check_is_rule_may_be_applied(rule, &current);
            path.push(rule_num);
            
            if can_apply {
                execute_rule(rule, &mut current, pos);
            }
        }
    }

    (current, path)
}


//будем генерировать 50000 тестов
//вернём true, если все тесты успешны,
//иначе false
//возвращаем строку, на которой была ошибка
//и последовательность правил
//ВАЖНО: 
//выбранную строку и строку, которая получится в исходной системе,
//доведём в новой системе до нормальных форм
//они должны совпасть 
fn fuzzing() -> (bool, (String, Vec<usize>)) {
    for _ in 0..50000 {
        let s = get_random_string(15);
        let (s2, path) = t(&s);
        
        let mut normalized_s = s.clone();
        let mut normalized_s2 = s2.clone();
        
        normalize(&mut normalized_s);
        normalize(&mut normalized_s2);

        if normalized_s != normalized_s2 {
            return (false, (normalized_s, path));
        }
    }

    (true, ("".to_string(), Vec::new()))
}

fn main() {
    init_rules_t();
    init_rules_t2();

    let (success, (error_string, path)) = fuzzing();

    if success {
        println!("все тесты успешны");
    } else {
        println!("найдена ошибка: {} и правила:", error_string);
        for rule_num in path {
            println!("{}", rule_num);
        }
    }
}